Issue#98 Allow user to specify locking to specific rev

Erich-McMillan opened the pull request:
Implements a mechanism through which the user may lock a specific external to a certain revision.

example:
```bash
gitman lock external1@rev1 group1 external2 # external1 will lock to rev1 other externals will lock to checked out rev
```

Erich-McMillan commented:
@jacebrowning this code implements the functionality specified by issue98, still writing unit tests since they seem to be broken on Windows.

Open questions:
- Should gitman verify that the revision actually exists in the repository, or just assume the user is correct?
- Should gitman allow the user to specify a revision pointed to by a tag or branch i.e. `gitman lock external1@path/to/tag` which would lock the external to the revision pointed to by the tag.

jacebrowning commented:
@Erich-McMillan for the initial version of this feature, I think we can assume that everything after the `@` is a valid value for `rev`, whether it a be a SHA, branch, or tag. This matches the experience of manually editing `gitman.yml` -- only when running `install`/`update` does one learn that the `rev` is invalid.

Erich-McMillan commented:
Jace, I'm unclear how the code could support non-SHA values in it's current form since that would break the sources_locked to have `rev:tag` or `rev:branch` right?

I have implemented the ability to use `git log` to determine whether a commit is valid and also enables this feature to accept tags/branches after the `@`. Should I split that out to a separate feature/PR or add it here?

jacebrowning commented:
> that would break the sources_locked to have rev:tag or rev:branch right?

Technically, `sources_locked` can use non-SHAs for certain use cases: https://gitman.readthedocs.io/en/latest/use-cases/linked-features/

But in that workflow, one can no longer run `gitman lock` or it will replace the branch with a SHA.

> I have implemented the ability to use git log to determine whether a commit is valid and also enables this feature to accept tags/branches after the @. Should I split that out to a separate feature/PR or add it here?

If it's straightforward to verify that `rev` is valid, feel free to include it here!

Erich-McMillan commented:
@jacebrowning I'm ready for feedback.

jacebrowning commented on gitman/models/source.py:
```diff
@@ -220,7 +220,7 @@ def identify(self, allow_dirty=True, allow_missing=True, skip_changes=False):
         raise self._invalid_repository
 
-    def lock(self, rev=None, allow_dirty=False, skip_changes=False):
+    def lock(self, rev=None, allow_dirty=False, skip_changes=False, verify_rev=True):
```
Is there any easy way to make this default to `False` to avoid potentially breaking changes?

Erich-McMillan replied:
Could add a top-level flag `--verify` and default to False if that's acceptable.

Erich-McMillan replied:
Though it will only verify *if* the user specifies the source in the format `sourcename@rev`, so I don't anticipate any breaking changes.

jacebrowning replied:
I don't think we need to go that far yet, but instead only pass `verify_rev=True` if the `rev` value from the CLI originally contained `@`, which was later parsed.

Erich-McMillan replied:
I think the point I was trying to make, was that when `rev==None`, then `verify_rev` will be ignored. Since the calling function `lock_dependencies` sets `rev=None` if `@` sign isn't present then as long as the user isn't using `sourcename@rev` it doesn't matter what `verify_rev` is set to.

I think maybe this behavior is confusing, I'm open to suggestions on how to improve it.

jacebrowning replied:
In that case, I think this implementation works!

jacebrowning commented on gitman/models/config.py:
```diff
@@ -180,11 +180,38 @@ def run_scripts(self, *names, depth=None, force=False, show_shell_stdout=False):
+        return true_name, rev
+
+    @classmethod
+    def remap_names_and_revs(cls, names):
```
If possible, I think these should be private methods.

Erich-McMillan replied:
By private you mean adding `__` to the beginning of the method name?

jacebrowning replied:
Just a single `_` to indicate to clients that these methods aren't meant for public consumption and might change.

jacebrowning commented on gitman/git.py:
```diff
@@ -259,6 +259,13 @@ def get_branch():
 
+def get_object_rev(object_name):
+    """Get the revision associated with the object specified."""
+    log_results = git("log", "-n", "1", object_name)
```
Would it make sense for this to capture exceptions and return a tuple of `commit_sha, valid` to be used by the caller? I think we want to avoid the CLI displaying exception stacktraces, which can happen now:

```
$ poetry run gitman lock gitman_1@main

Locking dependencies...

$ cd /Users/Browning/Code/gitman/demo

  $ cd gitman_1
  $ git log -n 1 main
  fatal: ambiguous argument 'main': unknown revision or path not in the working tree.
  Use '--' to separate paths from revisions, like this:
  'git <command> [<revision>...] -- [<file>...]'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/Browning/Code/gitman/gitman/cli.py", line 239, in main
    _run_command(function, args, kwargs)
  File "/Users/Browning/Code/gitman/gitman/cli.py", line 325, in _run_command
    success = function(*args, **kwargs)
  File "/Users/Browning/Code/gitman/gitman/decorators.py", line 9, in wrapped
    result = function(*args, **kwargs)
  File "/Users/Browning/Code/gitman/gitman/commands.py", line 264, in lock
    count = config.lock_dependencies(*names, obey_existing=False)
  File "/Users/Browning/Code/gitman/gitman/models/config.py", line 230, in lock_dependencies
    source_locked = source.lock(skip_changes=skip_changes, rev=rev)
  File "/Users/Browning/Code/gitman/gitman/models/source.py", line 236, in lock
    rev = git.get_object_rev(rev)
  File "/Users/Browning/Code/gitman/gitman/git.py", line 264, in get_object_rev
    log_results = git("log", "-n", "1", object_name)
  File "/Users/Browning/Code/gitman/gitman/git.py", line 16, in git
    return call("git", *args, **kwargs)
  File "/Users/Browning/Code/gitman/gitman/shell.py", line 94, in call
    raise ShellError(message, program=program, output=complete_output)
gitman.exceptions.ShellError: An external program call failed.

In working directory: /Users/Browning/Code/gitman/demo/gitman_1

The following command produced a non-zero return code:

$ git log -n 1 mainfatal: ambiguous argument 'main': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

Erich-McMillan replied:
Agreed, this should catch ShellErrors, log the error and then return `None` rather than a tuple.

jacebrowning replied:
Passing `_ignore=True` should handle this for you: https://github.com/jacebrowning/gitman/blob/c00f7c698640d6f4eef848c592321b484a81b1cd/gitman/shell.py#L15-L25

Erich-McMillan commented:
@jacebrowning are you going to create a new release for this feature?

jacebrowning commented:
@Erich-McMillan Here's an alpha release for testing: https://pypi.org/project/gitman/3.2a3/

jacebrowning commented:
@Erich-McMillan that's great to hear! I'll make a beta release to open this up to a wider audience.

Do you still have a need for the `add` command described [here](https://github.com/jacebrowning/gitman/issues/98#issuecomment-1059544228)?

Erich-McMillan commented:
> @Erich-McMillan that's great to hear! I'll make a beta release to open this up to a wider audience.
> 
> Do you still have a need for the `add` command described [here](https://github.com/jacebrowning/gitman/issues/98#issuecomment-1059544228)?

Hey Jace, no we don't need `add`. Noticed you moved it to a new feature request. Though I think it should be pretty straightforward to implement now.
